pkg/nftables/nftables.go

```diff
@@ -0,0 +1,204 @@
+package nftables
+
+import (
+	"errors"
+	"fmt"
+	"log/slog"
+	"net"
+
+	"github.com/google/nftables"
+	"github.com/google/nftables/expr"
+	"golang.org/x/sys/unix"
+)
+
+var Accept = nftables.ChainPolicyAccept
+
+const (
+	NatTable  = "kube_vip"
+	SNatChain = "kube_vip_snat"
+	DNatChain = "kube_vip_dnat"
+)
+
+func ApplySNAT(podIP, vipIP string, ignoreCIDR []string, IPv6 bool) error {
+
+	conn, err := nftables.New()
+	if err != nil {
+		panic(err)
+	}
+
+	err = Flush(conn, IPv6) // TODO: pass in the type of table we need to concern ourselves with
+	if err != nil {
+		panic(err)
+	}
+
+	rule, err := CreateRule(podIP, vipIP, ignoreCIDR, conn, IPv6)
+	if err != nil {
+		panic(err)
+	}
+	slog.Debug("[egress]", "table", rule.Table.Name, "chain", rule.Chain.Name, "expr", rule.Exprs)
+	conn.AddRule(rule) // Add the rule
+	conn.Flush()       // Commit the rule to nftables
+	err = conn.CloseLasting()
+	if err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+func GetTable(IPv6 bool) *nftables.Table {
+	// Default to IPv4
+	table := &nftables.Table{
+		Family: nftables.TableFamilyIPv4,
+		Name:   NatTable,
+	}
+
+	// Move to IPv6 if needed
+	if IPv6 == true {
+		table.Family = nftables.TableFamilyIPv6
+	}
+	return table
+}
+
+func GetSNatChain(IPv6 bool) *nftables.Chain {
+	policy := nftables.ChainPolicyAccept
+	return &nftables.Chain{
+		Name:     SNatChain,
+		Table:    GetTable(IPv6),
+		Type:     nftables.ChainTypeNAT,
+		Hooknum:  nftables.ChainHookPostrouting,
+		Priority: nftables.ChainPriorityNATSource,
+		Policy:   &policy,
+	}
+}
+
+func FilterTable(conn *nftables.Conn, tableName string) (table *nftables.Table, err error) {
+	tableList, err := conn.ListTables()
+	if err != nil {
+		return nil, fmt.Errorf("list tables %v", err)
+	}
+	for _, t := range tableList {
+		if t.Name == tableName {
+			return t, nil
+		}
+	}
+	return nil, nil
+}
+
+func Flush(conn *nftables.Conn, IPv6 bool) error {
+	if t, err := FilterTable(conn, NatTable); err != nil {
+		return err
+	} else if t != nil {
+		conn.DelTable(t)
+	}
+
+	// These don't return errors, so not 100% sure how to guarentee things were created
+	conn.AddTable(GetTable(IPv6))
+	conn.AddChain(GetSNatChain(IPv6))
+	return conn.Flush()
+}
+
+func CreateRule(podIP, vipIP string, ignoreCIDR []string, conn *nftables.Conn, IPv6 bool) (*nftables.Rule, error) {
+
+	// Get the kube-vip table
+	table := GetTable(IPv6)
+
+	// Create our rule
+	rule := &nftables.Rule{
+		Table: table,
+		Exprs: []expr.Any{},
+	}
+	// Set the correct chain
+	rule.Chain = GetSNatChain(IPv6)
+
+	// original pod IP
+	if net.ParseIP(podIP) == nil {
+		return nil, errors.New("ip is invalid")
+	}
+
+	// Create an element using our pod IP
+	elements := []nftables.SetElement{
+		{
+			Key: net.ParseIP(podIP).To4(),
```

Comment:
shouldn´t this reflect the IPv6 bool flag, instead of doing a To4 directly?

Reply:
fixed